### PR TITLE
[#IOCIT-301]  Add JWT management + key rotation management

### PR DESCRIPTION
**Pull request by gquadrati**
depends on https://github.com/pagopa/io-functions-lollipop/pull/15

#### List of Changes

Add JWT management with key rotation management

- [[#IOCIT-301] add new libriaries for jwt management](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/e2ef7a5c2e545fd6b427f39244277c2acf51cbb1)
- [[#IOCIT-301] add new env variables](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/382c3381d2c1af1844955a1116581fc3d4bd6404)
- [[#IOCIT-301] add JWT generator/validator](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/f1bcb4e0a99cbce9789d9d8d006e4b1a689f73c1)
- [[#IOCIT-301] add tests](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/cf80aba422459e150f6a82bb09791a2b43cfd580)

#### Motivation and Context

Enable LolliPoP Protocol

#### How Has This Been Tested?

Unit tests

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: utils/jwt_with_key_rotation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+export const getValidateJWT: GetValidateJWT = (
+  issuer,
+  primaryPublicKey,
+  secondaryPublicKey
+) => (token): TE.TaskEither<Error, jwt.JwtPayload> =>
+  pipe(
+    token,
+    validateJWTWithKey(issuer, primaryPublicKey),
+    TE.orElse(err =>
```
Comment (fabriziopapi) [inline utils/jwt_with_key_rotation.ts:71]: Could make sense to proceed only if the error is a JsonWebTokenError - invalid signature ?

File: utils/auth_jwt.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  assertionRef: AssertionRef,
+  operationId: OperationId
+});
+
+export type DecodedAuthJWT = jwt.JwtPayload & AuthJWT;
+
+/**
+ * AuthJWT Generation
+ */
+export type GenerateAuthGWT = (
```
Comment (AleDore) [inline utils/auth_jwt.ts:36]: typo? `GWT` -> `JWT`

File: utils/auth_jwt.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+) => TE.TaskEither<Error, DecodedAuthJWT>;
+
+export const getValidateAuthJWT = ({
+  ISSUER,
+  PRIMARY_PUBLIC_KEY,
+  SECONDARY_PUBLIC_KEY
+}: IConfig): ValidateAuthGWT =>
+  pipe(
+    getValidateJWT(ISSUER, PRIMARY_PUBLIC_KEY, SECONDARY_PUBLIC_KEY),
+    validateJWTFunction => (token): ReturnType<ValidateAuthGWT> =>
```
Comment (AleDore) [inline utils/auth_jwt.ts:64]: Do we need this typing here?

File: utils/config.ts [hunk 4]
```diff
@@ -16,15 +16,25 @@ import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
...
+    // JWT Support
+    ISSUER: NonEmptyString,
+    PRIMARY_PRIVATE_KEY: NonEmptyString,
+    PRIMARY_PUBLIC_KEY: NonEmptyString,
+
+    isProduction: t.boolean
+  }),
+  t.partial({
+    SECONDARY_PUBLIC_KEY: NonEmptyString
+  })
```
Comment (AleDore) [inline utils/config.ts:36]: Can we extract the intersection of the JWT Configuration type ?

File: utils/jwt_with_key_rotation.ts (hunk 1 above)
Comment (AleDore) [reply utils/jwt_with_key_rotation.ts:71]: I agree with @fabriziopapi

File: utils/auth_jwt.ts (hunk 3 above)
Comment (gquadrati) [reply utils/auth_jwt.ts:64]: Just to avoi putting `// eslint-disable-next-line @typescript-eslint/explicit-function-return-type` on top of it

File: utils/__tests__/jwt_with_key_rotation.test.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  });
+});
+
+describe("getValidateJWT - Success", () => {
+  it("should succeed validating a valid JWT generated with primary key, during standard period", async () => {
+    // Setup
+    const generateJWT = getGenerateJWT(issuer, aPrimaryKey.privateKey);
+    const token = await generateJWT(aPayload, aTtl)();
+
+    if (E.isRight(token)) {
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply utils/__tests__/jwt_with_key_rotation.test.ts:37]: Right :D  Done in [[#IOCIT-301] fix typos](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/42cf2bf39254eb367c61ac5b23961626d2b4b379)

File: utils/__tests__/jwt_with_key_rotation.test.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+      expect(result).toMatchObject(
+        E.left(E.toError("JsonWebTokenError - invalid signature"))
+      );
+    } else {
+      fail("Invalid token");
+    }
+  });
+
+  it("should fail validating a invalid JWT with both public primary key and public secondary key, during key rotation period", async () => {
```
[the remark this answers is not in the document]
Comment (gquadrati) [reply utils/__tests__/jwt_with_key_rotation.test.ts]: Done in [[#IOCIT-301] fix typos](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/42cf2bf39254eb367c61ac5b23961626d2b4b379)

File: utils/config.ts (hunk 4 above)
Comment (gquadrati) [reply utils/config.ts]: Done in [[#IOCIT-301] refactor IConfig + TTL from config](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/431284c74584bf8c56a106fcbc8cef8d057e3cf4) along with the move of the JWT_TTL on config (with fallback to 900 sec)

File: utils/jwt_with_key_rotation.ts (hunk 1 above)
Comment (gquadrati) [reply utils/jwt_with_key_rotation.ts:74]: Right. Done in [[#IOCIT-301] second validation when error is invalid signature](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/6105c1d9a7905b5235e4941665fcff499fc8ac77)

File: utils/auth_jwt.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+ * Type Definitions
+ */
+
+export type AuthJWT = t.TypeOf<typeof AuthJWT>;
+export const AuthJWT = t.interface({
+  assertionRef: AssertionRef,
+  operationId: OperationId
+});
+
+export type DecodedAuthJWT = AuthJWT;
```
Comment (AleDore) [inline utils/auth_jwt.ts:27]: Do we need this?

File: utils/jwt_with_key_rotation.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+export type GetGenerateJWT = <T extends Record<string, unknown>>(
+  issuer: NonEmptyString,
+  primaryPrivateKey: NonEmptyString
+) => (payload: T, ttl: Second) => TE.TaskEither<Error, NonEmptyString>;
+
+export const getGenerateJWT: GetGenerateJWT = (
+  issuer,
+  primaryPrivateKey
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+) => (payload, ttl) =>
```
Comment (AleDore) [inline utils/jwt_with_key_rotation.ts:33]: Can we explicit return type here?

Comment (gquadrati) [reply utils/jwt_with_key_rotation.ts]: Done in [[#IOCIT-301] remove DecodedAuthJWT + minor fix](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/915afe8d302a5022519a7819395f187c71a9c485)

File: utils/auth_jwt.ts (hunk 7 above)
Comment (gquadrati) [reply utils/auth_jwt.ts]: Actually, not anymore. Removed in [[#IOCIT-301] remove DecodedAuthJWT + minor fix](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/915afe8d302a5022519a7819395f187c71a9c485)

File: utils/__tests__/auth_jwt.test.ts [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+  })
+);
+
+describe("getGenerateJWT", () => {
+  it("should generate a valid AuthJWT", async () => {
+    const generateJWT = getGenerateAuthJWT(aConfigWithPrimaryKey);
+
+    const res = await generateJWT(aPayload)();
+
+    console.log(res);
```
Comment (Garma00) [inline utils/__tests__/auth_jwt.test.ts:41]: Do we need this?

File: utils/jwt_with_key_rotation.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  issuer: NonEmptyString,
+  key: NonEmptyString
+) => (token: NonEmptyString) => TE.TaskEither<Error, jwt.JwtPayload> = (
+  issuer,
+  key
+) => (token): TE.TaskEither<Error, jwt.JwtPayload> =>
+  pipe(
+    TE.tryCatch(
+      () =>
+        new Promise<jwt.JwtPayload>((verify, reject) => {
```
Comment (AleDore) [inline utils/jwt_with_key_rotation.ts:97]: why `verify` instead of `resolve` ?

Comment (gquadrati) [reply utils/jwt_with_key_rotation.ts:97]: Just a typo :) Renamed in [3195e23](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/3195e237f97d2bc6a9790be6394843bca79bdae4)

File: utils/__tests__/auth_jwt.test.ts (hunk 9 above)
Comment (gquadrati) [reply utils/__tests__/auth_jwt.test.ts:41]: Good catch! Removed in [3195e23](https://github.com/pagopa/io-functions-lollipop/pull/14/commits/3195e237f97d2bc6a9790be6394843bca79bdae4)

**Review by Garma00 — APPROVED** [review]

**Review by AleDore — APPROVED** [review]

**Pull request merged**